### DISCUSSION TITLE
Hypothetical wishlist ideas

### DISCUSSION THREAD

**Discussion by markerikson** [reduxjs/redux-toolkit#1847] [Ideas]
Opening this up for random ideas that we could at least _consider_ adding to RTK or the Redux ecosystem someday.  I'm creating this as a discussion rather than an issue because A) better threading, and B) won't annoy me if this never gets closed

Starting off, from [a Reddit thread I bookmarked a couple years ago](https://www.reddit.com/r/javascript/comments/bja4wr/application_state_management_with_react_by_kent_c/em9wln4/?context=3):

1.

> Of applications that do need redux, he's completely on the money about them getting progressively more complicated as redux is leveraged for prop drilling more than global state management.

> One of the things I'm still astonished by with redux is the lack analysis tools that can examine your state management, generate a report, and show you all of the dark crevices a state property winds up and how many side effects a single action can have.

2.

> As far as I've been able to tell there isn't a documentation tool that shows where an action can get fired from. What middleware is attending to it (and what actions it fires off as a result), and finally what reducer listens to it.

> Showing how data is changed would be awesome (but not really practical).. however just showing all of the places an action has side-effects against in a flow chart or process diagram would be a big win for just touching an enterprise redux code base without breaking gobs of stuff as a result because of lack of awareness.

> Redux dev tools doesn't seem like it's really met that requirement yet... and waiting until the action actually runs doesn't make for good static documentation. I figure something like this could inspect reducers and middleware during unit tests and generate a report after the unit tests execute...

> might also be useful for determining coverage for all of your known actions (assuming you have some defined list of action types).

**Comment by nathggns** [comment]
Something to do with data driven side effects - I.e, when the state matches X condition (probably driven by a selector), do Y, until it no longer matches that condition, in which case it's cancelled. Think for example, a music player where there is a selector that indicates if it should be playing or not. We can drive this by actions with things like thunk/sagas, but definitely becomes difficult if there's a million different actions that can change this state.

**Reply by markerikson** [reply]
My first response is that sounds like your best bet is mashing together XState and Redux, by using an XState machine as a Redux reducer and hooking the service up as a middleware per this example:

https://github.com/mattpocock/redux-xstate-poc

**Comment by Faithfinder** [comment]
Well, personally I think one of the worst parts of redux experience is action dispatching. I understand that it's largely what makes defines redux, and I know that a hook analog to mapDispatchToProps was thought about and discarded, but I still believe that this part could have better DX. Somehow. (no specific ideas, sorry)

**Reply by markerikson** [reply]
Not quite sure what you're describing here. Can you give an example?  Is it about needing to write `const onClick = () => dispatch(increment())` manually in your component with `useDispatch()`, vs being able to do `props.increment()` with `connect()`?

**Reply by mikoz93** [reply]
That would be really nice. Not having to pull in `useDispatch` in the components/hooks we need to dispatch actions from and instead just directly calling the actions created by the `createSlice` / `createAction`.

**Reply by markerikson** [reply]
Yeah, that's not feasible, I'm afraid.

React-Redux has _always_ relied on reading the current Redux store instance via context.  Both `connect`, `useDispatch`, and `useSelector` all rely on accessing that same context-provided store instance to read state and dispatch actions.  That's why those APIs exist in the first place.  

The action creator functions in your `counterSlice.ts` file don't have any idea about a "store" at all. They're just plain functions that return an action object.  They only have meaning when you call them and then pass the resulting action object to `dispatch()`, so you have to have access to `dispatch` at the time you want to use it.

The idea of having an automatic wrapper for binding action creators is a separate question.  We dropped that during the React-Redux hooks alpha development, because A) it was clear there'd be name clashes between imported action creators at the top of a file and bound versions of those action creators inside the component, and B) Dan suggested it was time to remove the idea of "binding action creators" as an abstraction in order to make it more clear what is actually going on (as that was a common complaint with `connect` being too "magical").

FWIW it's pretty easy to write a `useBindActionCreators` hook or something that wraps `useDispatch`, but at this point I don't ever seeing us add that as an actual published API to React-Redux (or RTK).

**Reply by mikoz93** [reply]
Regarding the name collisions, that is resolved quite easily with a namespace import such as `import * as todoActions from '...'`. This is a problem that already exists today and is already being solved that way. Regarding the `magic`, I think this would take probably about a sentence or two in the docs to explain how it works, and in turn we could save ourselves writing extra code that is meant to work together anyway. I'd be tempted to say that the magic of something like this would be nothing compared to how magical `createSlice` feels.

**Reply by Faithfinder** [reply]
> Not quite sure what you're describing here. Can you give an example? Is it about needing to write `const onClick = () => dispatch(increment())` manually in your component with `useDispatch()`, vs being able to do `props.increment()` with `connect()`?

Indeed! And while I _do_ talk about all actions, it is especially felt with `thunk` actions - when you describe a typical simple thunk, i.e

```ts
export const getUser = (id: string) => async (dispatch: AppDispatch) => {
    const user = await client.get(`/user/${id}`);
    dispatch(setUser(user));
};
```
Or maybe with the `createAsyncThunk`
```ts
export const fetchUser = createAsyncThunk('users/fetchUser', async () => {
  const user = await client.get(`/user/${id}`);
  return user;
})
```

The thunk really looks like it takes care of everything, but it doesn't and you still need to remember to dispatch from the component. It just feels unnatural. I don't need an explanation of how and why that it is like it is - I understand.  I just wish for some magic to make it feel better.
I guess part of the answer is just "use RTKQ"? But I bet it's not suitable for all cases.

And I agree with Mikoz that whatever magic gets introduced is nothing compared to the rest of the abstractions from the toolkit. Like, if you do go RTKQ way you're barely even aware that Redux is involved.

**Reply by markerikson** [reply]
@Faithfinder I guess I'm still not quite clear on what specific problem you're pointing to here.  Is it that Redux requires you to `dispatch()` something in order to do anything at all, or the fact that you have to explicitly reference the `dispatch()` function in your own code?

**Reply by Faithfinder** [reply]
> @Faithfinder I guess I'm still not quite clear on what specific problem you're pointing to here. Is it that Redux requires you to `dispatch()` something in order to do anything at all, or the fact that you have to explicitly reference the `dispatch()` function in your own code?

Aren't those the same problem? Or can you do redux without dispatching? But yes, I mean explicit referencing in my own code.

**Reply by markerikson** [reply]
No, those are two different things: the need to `dispatch()` to make something happen, vs whether the actual function name `dispatch` shows up in the code that you write.

Redux _requires_ dispatching. Period. That is how the Flux Architecture works. Literally the only way to trigger update an update to the store is to dispatch an action object.  Since [thunks are a way to write code that will interact with the store, without knowing ahead of time _which_ store they talk to](https://redux.js.org/usage/writing-logic-thunks#why-use-thunks), they _also_ require passing the thunk function to `dispatch`.  So, all that behavior is non-negotiable, because without it it's no longer Redux.

The question of whether you yourself literally have to write `dispatch()` in your code is different.  `connect()` intentionally abstracted that, so that your wrapped component would be able to call `props.increment()` and have that be a pre-bound action creator that calls `dispatch()` internally, without your own component needing to "know" that it was actually dispatching.

Now, that wasn't strictly _necessary_.  After all, if you didn't provide a `mapDispatch` argument, your component automatically got `this.props.dispatch` instead, and you could always do `this.props.dispatch(increment())` in the same way we now do it with `useDispatch()`.  But, our encouraged pattern at the time was to use `mapDispatch` so that the whole "container/presentational" thing was used and the wrapped component was, in theory, more reusable and testable.

With hooks, that calculus changes.  As I described in my post [Thoughts on React Hooks, Redux, and Separation of Concerns](https://blog.isquaredsoftware.com/2019/07/blogged-answers-thoughts-on-hooks/) and my talk [ReactBoston 2019: Hooks, HOCs, and Tradeoffs](https://blog.isquaredsoftware.com/2019/09/presentation-hooks-hocs-tradeoffs/), hooks change the way we write components so that they are now responsible for dealing with their own data dependencies.  That includes Redux.  

So, between the technical problems of potential name clashes with a `useActions` hook, the removal of the need to pass bound action creators as props, and a desire to remove a layer of abstraction that is no longer necessary and often confused people, we settled on just providing `useDispatch()` and telling users "write your own click handlers that call `dispatch()` when you need to".

Yeah, it's a couple more lines of code. Like anything else, _you_ can write your own abstractions for this if you really want to. 

But at this point it's not a problem that needs to be solved further.

**Reply by mikoz93** [reply]
> Like anything else, you can write your own abstractions for this if you really want to.

Isn't the whole reason behind RTK is to simplify already existing solutions and take away the need for people to make their own abstractions? I think what's being suggested here is to go a step further with the current abstractions provided by RTK, such as `createAction` and wrap the `dispatch` in whatever is being returned. Consumers don't really care what it does under the hood as long as it provides the same interface (minus the manual dispatch of course).

**Reply by markerikson** [reply]
I'm saying that A) I don't think this is a problem that _needs_ to be solved, B) it's not a problem that I _want_ to try to address given the previous discussions during the hooks alpha development.

To be clear, this is really a React-Redux-ish question. Yes, we do have some React-based APIs included in RTK now that we have RTK Query, but there's nothing about the existing RTK APIs that would be relevant here. The only vaguely viable technical option that matches what I _think_ you folks are describing is the `useActions()` hook that we already discarded during the hooks alpha. That would be more of a React-Redux thing than an RTK thing, but I truly don't see that as a pain point that's worth adding a new API for.

**Reply by phryneas** [reply]
I'll add to this:

`dispatch` has to come from *somewhere*.  

We cannot do that with a fixed store binding during the `createSlice` call, since at that point no store exists yet. Also, we would not do it there, since a fixed binding at that point would greatly hinder many types of unit and integration tests that rely on the existance of multiple stores.  

We could provide `useActions()`, but that hook has been a copy-paste solution as part of the Redux hooks, (if you wanted to, you could already be using it for years) and it has the problems Mark already added.  
If you were to use that, it would look like
```js
const { increment, decrement } = useActions(counterActions)
```
At that point it is incredibly difficult to explain to a learner why they could not just call `counterActions.increment()` instead. Even worse, you get shadowing if you do not do a `import * as counterActions from './counterSlice'`, but import them manually. Also, it gets even more error-prone that way. Is `increment` already bound or not? That's confusing.  
And we *want* people to import individual actions. Because that works well with IDE autocomplete. They don't even need to know which file it comes from (and going by [handle actions as events, not setters](https://redux.js.org/style-guide/style-guide/#model-actions-as-events-not-setters), they should not care what file it comes from - they are dispatching an action that could be handled in 20 different reducers at that point), they just write `increm` and hit `ctrl + space`.  

Then there would be that last solution of creating one hook per action creator. But does that really save any code?
```js
const increment = useIncrementAction()
const decrement = useDecrementAction()

return <><button onClick={increment}>+</button><button onClick={decrement}>-</button></>
```
vs
```js
const dispatch = useDispatch()

return <><button onClick={() =>dispatch(increment)}>+</button><button onClick={() => dispatch(decrement)}>-</button></>
```
This approach in my eyes does not save any code. It just adds magic and hides intent.

**Comment by MikelArnaiz** [comment]
I'd like to see built in code splitting support in RTK. And I'd love to know your opinion of how this should look like.
I know about the [The problem of code splitting](https://github.com/reduxjs/redux/issues/4010) (mainly SSR) but there are some options out there, like [the suggestions from the docs](https://redux.js.org/usage/code-splitting) or [the reducer registry](https://nicolasgallagher.com/redux-modules-and-code-splitting/). 

Regarding libraries offering just that ([redux-dynostore](https://github.com/ioof-holdings/redux-dynostore), [redux-injectors](https://github.com/react-boilerplate/redux-injectors), [redux-dynamic-modules](https://github.com/Microsoft/redux-dynamic-modules)) they are all, unfortunately, deprecated or semiabandoned.

An out of the box solutions will probably encourage its use.

**Reply by markerikson** [reply]
Yeah, the most recent one that popped up just a few days ago is https://github.com/fostyfost/redux-eggs .  

I think this is a use case we could potentially try to address, but I'd say this is a lower priority on our end atm.
